README.md
## DNS Client library with no dependencies ##
```libdns``` is my 're-invention of the wheel' DNS client resolver
library, written entirely in C with no external dependencies.

Supports UDP as transport protocol only (Meaning, no Zone Transfers).

I even wrote a man page.

Supplied with an example client application: ```dnsq```.

### Do note... ###
This code is presented nearly as-is from when it was written back
in 2003. The world and my abilities have come a long way in over
16 years...


HISTORY

  Update History  -  DNS Library
  ------------------------------

6th September, 2003 - Implemented timeouts for dns_send_query() using
    	       	      select(). Note that the actual call to select() is
		      made in dns_get_response_udp().
		    - Did a test build on Mac OS X, 10.3 and aside from
		      differing linker flags to make shared objects, things
		      went without a hitch. This also prompted the use of
		      <stdint.h> to declare the 'uint32_t' etc. data types.

2nd September, 2003 - Fixed recursion problem with dns_brs_helper(). Turns
    	       	      out I was only using the bottom byte of a compression
		      word as the offset instead of the LS 14 bits. Duh!

31st August, 2003   - Replaced occurances of 'unsigned short', 'unsigned int'
     	     	      and 'unsigned long' with portable types, 'uint16_t'
		      and 'uint32_t'. Note that some int's were left as they
		      are. This happens where the integer's size and sign
		      isn't important (eg. flag values).
		    - Modified Makefile to create static library and install
		      libraries with versioned symbolic links.

30th August, 2003   - *Partially* fixed dns_build_reply_string(). At least
      	     	      implemented a recursive method that seems to handle
		      most compressed strings fine. Some strings still give
		      it problems though. At least now the routine uses
		      recursion.

27th August, 2003   - Extended reply handling to actually parse what
(only just the 27th)  is returned and add it to relevant lists (well, only
		      answers are implemented so far).
		    - Added utility symbol-to-string conversion routines
		      for types, classes and error codes.
27th August, 2003   - Implemented dns_build_rr() to remove redundancy in
(almost the 28th...)  dns_parse_response().


dns.c
/* DNS Query construction and response parsing code.
 * 
 * Written by Owen Klan  -  23rd August 2003
 */

/* TODO:
 * - review usage of data pointer of reply packet. Be careful as to 
 * where it is free()'d, I have a suspicion of a memory leak... */

#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <errno.h>

#include <netinet/in.h>
#include <sys/socket.h>

#include "dns.h"


/* Types defined for dns_add_to_list(). These definitions can stay
 * internal to dns.c without issues. */
#define DNS_QUESTION    1              /* Question structure */
#define DNS_RR          2              /* Resource Record structure */

/*
 * Static function prototypes or external prototypes that library users
 * don't need to see. */
static int dns_add_to_list(dnsq_t *q, void **list, void *obj, int type);
static int dns_parse_response(dnsq_t *q, char *data, int len);
static struct dns_rr *dns_build_rr(dnsq_t *, char *data, char *start,
                                   int len, int *dist);

/* From dns_udp.c */
extern int dns_send_query_udp(dnsq_t *q, int s, struct sockaddr_in *addr);
extern int dns_get_response_udp(dnsq_t *q, int s, struct sockaddr_in *addr,
                                int timeout, char **data, int *len);

/* From dns_str.c */
//extern char *dns_build_reply_string(char *data, int len, char *start,
//                                  int *dist);

/* This function will add a question record to a previously allocated
 * DNS Query structure. Returns 0 on success, or one of the error codes
 * defined in dns.h.
 * 
 * Parameters:
 *    q :          DNS Query structure previously allocated
 *    host :       The hostname (or IP) to query
 *    type :       Query type (see dns.h)
 *    class:       Query class (see dns.h)
 */
int dns_add_question(dnsq_t *q, char *host, uint16_t type,
                     uint16_t class)
{
    struct dns_question *quest;
    int err = 0;
    
    if (!q || !host)                   /* Check pointer validity */
      return DNS_ERR_NULL_PARAM;
    
    if ((quest = malloc(sizeof(struct dns_question))) == NULL) {
        q->dns_errno = errno;
        return DNS_ERR_SEE_ERRNO;              /* Most likely memory alloc. failure */
    }
    memset(quest, 0x00, sizeof(struct dns_question));
    
    /* Convert the requested hostname to the required DNS format */
    if ((quest->q_content = dns_convert_name(host, &err)) == NULL) {
        free(quest);
        if (err == DNS_ERR_SEE_ERRNO)    /* Save errno if necessary */
          q->dns_errno = errno;
        return err;
    }
    quest->q_type = type;              /* Setup type, class and next ptr. */
    quest->q_class = class;
    quest->next  = NULL;
    
    /* Now add this item to the list */
    dns_add_to_list(q, (void *)&q->questions, (void *)quest, DNS_QUESTION);

    q->header.dns_num_quest++;         /* One more question to pile... */
    
    return 0;
}


/* This function returns a new DNS Query structure, ready to have
 * questions added to it. Returns NULL on error. */
dnsq_t *dns_new_query(uint16_t id, int opcode, int recursive)
{
    dnsq_t *q;                         /* Query structure */
    
    if ((q = malloc(sizeof(dnsq_t))) == NULL) {
        return NULL;
    }
    
    /* Zero out the entire structure */
    memset(q, 0x00, sizeof(dnsq_t));
    
    /* Now setup initial fields */
    q->header.dns_id = id;
    q->header.dns_flags = 0;
    DNS_SET_FLAGS(q, opcode, recursive);
    
    return q;
}

/* This procedure will destroy *ALL* memory associated with a DNS Query
 * structure. It will free the lists of questions, all RRs and any data
 * that was saved as part of an RR. Finally, it will free the query
 * structure itself */
void dns_free_query(dnsq_t *q)
{
    struct dns_question *quest, *quest_old;
    struct dns_rr *rr, *rr_old;
    
    /* Free the lists of questions and RRs first */
    quest = q->questions;
    
    while (quest) {
        FREE_IF_VALID(quest->q_content);   /* Free content string */
        quest_old = quest;
        quest = quest->next;               /* Next element */
        free(quest_old);                   /* Free old element */
    }
    
    rr = q->answers;                   /* Free the answers list */
    while (rr) {
        FREE_IF_VALID(rr->rr_name);
        FREE_IF_VALID(rr->rr_data);
        rr_old = rr;
        rr = rr->next;
        FREE_IF_VALID(rr_old);
    }

    rr = q->authority;                 /* Free the authority list */
    while (rr) {
        FREE_IF_VALID(rr->rr_name);
        FREE_IF_VALID(rr->rr_data);
        rr_old = rr;
        rr = rr->next;
        free(rr_old);
    }

    rr = q->additional;                /* Free the additional list */
    while (rr) {
        FREE_IF_VALID(rr->rr_name);
        FREE_IF_VALID(rr->rr_data);
        rr_old = rr;
        rr = rr->next;
        free(rr_old);
    }

    /* Free the reply data */
    FREE_IF_VALID(q->dns_data);
    
    free(q);                           /* Done */
}

/* Function that sends off a DNS request. Requires a sockaddr_in structure
 * specifying the destination server and the DNS query to send. Will return
 * zero on success, other wise an error code is returned. */
int dns_send_query(dnsq_t *q, int s, struct sockaddr_in *addr, int timeout)
{
    int retval;
    char *response_data;
    int data_size = 0;                 /* Hold size of returned data */
    
    if (!q || !addr)                   /* Confirm pointer validity */
      return DNS_ERR_NULL_PARAM;
    
    /* 23rd August 2003:  Currently only UDP transmission is supported! */
    if ((retval = dns_send_query_udp(q, s, addr)) > 0)
      return retval;
    if ((retval = dns_get_response_udp(q, s, addr, timeout, &response_data,
                                       &data_size)) > 0)
      return retval;

    retval = dns_parse_response(q, response_data, data_size);
    free(response_data);
    return retval;
}

/* Function that parses the DNS response data into lists of records
 * and also performs some flags checks. Returns zero on success,
 * error code on failure.
 * 
 * Parameters:
 *      q    Pointer to DNS query structure to work with
 *   data    Pointer to buffer containing reply data
 *    len    Length of returned reply data 
 */
static int dns_parse_response(dnsq_t *q, char *data, int len)
{
    struct dns_rr *rr;                 /* Response record */
    int i;
    int ret_val;
    int reply_code;
    int offset = sizeof(struct dns_header);      /* Offset in data */
    
    /* Check the truncated response flag first */
    if (q->header.dns_flags & 0x0200)
      return DNS_ERR_REPLY_TRUNCATED;
    
    /* Now check the reply code */
    reply_code = q->header.dns_flags & 0x000F;

    /* TODO:  Properly implement these checks */
    /* Note that a root server will most likely ignore queries that
     * request recursive resolution. */
    switch (reply_code) {
      case DNS_R_FORMAT_ERROR:         /* Format error with sent query */
        FREE_IF_VALID(data);
        return DNS_ERR_BAD_FORMAT;

      case DNS_R_NAME_ERROR:           /* No such name exists */
        FREE_IF_VALID(data);
        return DNS_ERR_NO_SUCH_NAME;

      case DNS_R_SERVER_FAILURE:       /* Server reports failure */
        FREE_IF_VALID(data);
        return DNS_ERR_SERVER_FAILURE;

      case DNS_R_NOT_IMPLEMENTED:      /* Requested query not supported */
        FREE_IF_VALID(data);
        return DNS_ERR_NOT_IMPLEMENTED;
        
      case DNS_R_REFUSED:              /* Server refused to handle query */
        FREE_IF_VALID(data);
        return DNS_ERR_QUERY_REFUSED;
    }
    
    /* So far all is fine. Now parse response into lists of items */
    /* Assume that the question section remains unchanged and start
     * with the answers section instead. */
    for (i = 0; i < q->header.dns_num_quest; i++) {
        offset += strlen((char *)(data + offset)) + 1;
        offset += 4;
    }

    /* Now offset is at beginning of Answers section */
    for (i = 0; i < q->header.dns_num_ans; i++) {
        int dist = 0;
        
        rr = dns_build_rr(q, data, (data + offset), len, &dist);
        
        /* Update the offset to point to the next record */
        offset += (dist + 10 + rr->rr_datalen);
        
        /* Add this to the list of answers */
        if ((ret_val = dns_add_to_list(q, (void *)&q->answers,
                                       (void *)rr, DNS_RR)) != 0) {
            /* Some error has occurred */
            FREE_IF_VALID(data);
            free(rr);
            return ret_val;
        }
    }
    
    /* Now offset is at beginning of Authority section */
    for (i = 0; i < q->header.dns_num_auth; i++) {
        int dist = 0;
        
        if ((rr = dns_build_rr(q, data, (data + offset),
                               len, &dist)) == NULL) {
            FREE_IF_VALID(data);
            free(rr);
            
            /* This is potentially nasty... */
            q->dns_errno = ENOMEM;
            return DNS_ERR_SEE_ERRNO;
        }
        
        /* Update the offset to point to the next record */
        offset += (dist + 10 + rr->rr_datalen);
        
        /* Add this to the list of answers */
        if ((ret_val = dns_add_to_list(q, (void *)&q->authority,
                                       (void *)rr, DNS_RR)) != 0) {
            /* Some error has occurred */
            FREE_IF_VALID(data);
            free(rr);
            return ret_val;
        }
    }

    /* Now offset is at beginning of Additional section */
    for (i = 0; i < q->header.dns_num_add; i++) {
        int dist = 0;
        
        if ((rr = dns_build_rr(q, data, (data + offset),
                               len, &dist)) == NULL) {
            FREE_IF_VALID(data);
            free(rr);
            
            /* This is potentially nasty... */
            q->dns_errno = ENOMEM;
            return DNS_ERR_SEE_ERRNO;
        }
        
        /* Update the offset to point to the next record */
        offset += (dist + 10 + rr->rr_datalen);
        
        /* Add this to the list of answers */
        if ((ret_val = dns_add_to_list(q, (void *)&q->additional,
                                       (void *)rr, DNS_RR)) != 0) {
            /* Some error has occurred */
            FREE_IF_VALID(data);
            free(rr);
            return ret_val;
        }
    }

    /* Duplicate the query data */
    if ((q->dns_data = malloc(len)) == NULL) {
        q->dns_errno = errno;
        return DNS_ERR_SEE_ERRNO;
    }
    memcpy(q->dns_data, data, len);
    
    return 0;
}


/* Function that builds a resource record. Returns a copy of the
 * dns_rr structure on sucess, NULL on failure.
 * 
 * Parameters:
 *   q         DNS query structure we are working with
 *   data      pointer to beginning of packet data
 *   start     pointer to beginning of RR
 *   len       length of data buffer (currently unused  27-08-2003)
 *   dist      address of integer to store the number of bytes this
 *             RR takes up in the reply.
 */
static struct dns_rr *dns_build_rr(dnsq_t *q, char *data, char *start,
                                   int len, int *dist)
{
    uint16_t rdl = 0;          /* Resource Data Length */
    struct dns_rr *rr;
    int rr_size = 0;                   /* Place to store bytes processed by
                                        dns_build_reply_string(). */
    
    if ((rr = malloc(sizeof(struct dns_rr))) == NULL) {
        /* Malloc failure. */
        return NULL;
    }
    
    rr->rr_name = dns_build_reply_string(data, len, start, &rr_size);
    rdl = ntohs(*(uint16_t *)(start + rr_size + 8));
    rr->rr_datalen = rdl;
    rr->rr_type    = ntohs(*(uint16_t *)(start + rr_size));
    rr->rr_class   = ntohs(*(uint16_t *)(start + rr_size + 2));
    rr->rr_ttl     = ntohl(*(uint32_t *)(start + rr_size + 4));
    
    /* Duplicate this resource records data section. */
    if ((rr->rr_data = malloc(rr->rr_datalen)) == NULL) {
        FREE_IF_VALID(rr->rr_name);
        FREE_IF_VALID(rr);
        return NULL;
    }
    
    memcpy(rr->rr_data, (start + rr_size + 10), rr->rr_datalen);
    rr->next    = NULL;
    
    *dist = rr_size;
    return rr;
}

/* Helper function that will add an item to a list in a dnsq_t structure.
 * Returns 0 on success, error code on error. */
static int dns_add_to_list(dnsq_t *q, void **list, void *obj, int type)
{
    if (!q || !list || !obj)           /* Check pointer validity */
      return DNS_ERR_NULL_PARAM;
    
    switch (type) {
      case DNS_QUESTION: {             /* Add to a question list */
          struct dns_question *current = (struct dns_question *)*list;
          
          if (!current) {              /* List is empty */
              *list = (struct dns_question *)obj;
          } else {
              while (current->next)    /* Add to end of list */
                current = current->next;
              
              current->next = (struct dns_question *)obj;
          }
          break;
      }
      case DNS_RR: {                   /* Add to a resource record list */
          struct dns_rr *current = (struct dns_rr *)*list;
          
          if (!current) {              /* List is empty */
              *list = (struct dns_rr *)obj;
          } else {
              while (current->next)    /* Add to end of list */
                current = current->next;
              
              current->next = (struct dns_rr *)obj;
          }
          break;
      }
      default:
        return DNS_ERR_PARAM_ERROR;            /* Unknown type */
    }
    
    return 0;
}


dns_str.c
/* String routines for DNS library. Split from dns.c on 28th August, 2003.
 * 
 * Written by Owen Klan  -  27th August, 2003
 */

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#include "dns.h"

/* This function will take an address in the form foo.bar.com and
 * convert it to the required format. For example:
 * 
 *           foo.bar.com  would give:   0x03foo0x03bar0x03com0x00
 * 
 * Will return NULL on error, otherwise returns a malloc()ed string
 * that the caller must free when finished with. The error parameter
 * is optional and if non-NULL, will be used to store an error code. */
char *dns_convert_name(char *name, int *error)
{
    char *conv_name = NULL;            /* Holds string to return to user */
    char *tok_start, *tok_end;         /* Used for tokenising */
    int len;                           /* Length of created string. */
    int tok_len;
    
    if (!name) {                       /* Make sure we have a name string */
        SET_ERROR(DNS_ERR_NULL_PARAM);
        return NULL;
    }
        
    tok_start = tok_end = name;
    len = strlen(name) + 2;
    
    if ((conv_name = malloc(len)) == NULL) {
        SET_ERROR(DNS_ERR_SEE_ERRNO);
        return NULL;
    } else {
        memset(conv_name, 0x00, len);
    }
    
    /* Tokenise the hostname into individual components. */
    for (; tok_end < (name + strlen(name));) {
        if ((tok_end = strchr(tok_start, '.')) == NULL) {
            /* Last token */
            tok_end = (char *)tok_start + strlen(tok_start);
        }
        
        tok_len = (int)tok_end - (int)tok_start;
        
        /* If the token length is > 63, then we will abort the
         * string construction. */
        if (tok_len > 63) {
            SET_ERROR(DNS_ERR_TOKEN_TOO_LONG);
            free(conv_name);
            return NULL;
        }
        
        /* Copy token length then token into constructed string */
        *(unsigned char *)(conv_name + strlen(conv_name)) =
          (unsigned char)tok_len;
        strncpy((conv_name + strlen(conv_name)), tok_start, tok_len);
        tok_start = tok_end + 1;
    }
    
    return conv_name;
}

/*
 * TODO:  Recheck this function when my mind is clearer...
 *  - make sure pointer calculations are as sane as possible.
 *  - *** CRITICAL ***  Fix recursion problem (re: hitting bottom of
 *    stack)
 */
/* 'count' is non-zero if the routine is just to count the length of a
 * generated string. 'level' is used as a recursion control to make sure
 * we count the moved offset properly. It is set to one only when this
 * function is called from outside. If we call ourselves, we make sure
 * level is zero.
 * 
 * Returns length of created string on success, -1 on error. */
static int dns_brs_helper(char *data, char *start, char *buffer,
                          int *dist, int count, int level)
{
    int len = 0;
    int tok_size = 0;
    uint16_t comp_offset = 0;
    int offset = 0;
    int top_level = level;
    int seg_size = 0;          /* Size of this string segment */
        
    do {    
        tok_size = *(unsigned char *)(start + offset);

        if (tok_size == 0) {           /* Eliminate last dot */
            *(buffer + strlen(buffer) - 1) = '\0';
            break;
        }

        /* Is this component compressed? */
        if (tok_size & 0xC0) {
            /* Get the offset into the reply data of the string
             * component in question. */

            comp_offset = (ntohs(*(unsigned short *)(start + offset)))
              & ~0xC000;
            
            dns_brs_helper(data, (data + comp_offset), buffer,
                           dist, count, 0);
            if (top_level) {
                seg_size += 2;
                *dist = seg_size; /* Return how far we moved from start */
            }
            return len;
        }
        
        /* Are we to actually copy data? */
        if (!count) {
            memcpy((buffer + strlen(buffer)), (start + offset + 1),
                    tok_size);

            /* Append dot if necessary */
            *(buffer + strlen(buffer)) = '.';
        }
        
        len += (tok_size + 1);

        if (top_level && tok_size > 0)
          seg_size += (tok_size + 1);

        offset += (tok_size + 1);
    } while (1);
    
    *dist = seg_size + 1;

    return len;
}

/* Function that will parse returned string values in replies and build
 * a full string, taking into account the DNS compression scheme. Returns
 * a malloc()'ed buffer with the string or NULL on failure.
 * 
 * data     is a pointer to the FULL reply packet data.
 * start    is a pointer (which resides within data) that the string in
 *          question begins at.
 * len      is the size of data buffer
 * dist     pointer to an integer location to store the number of bytes
 *          moved /RELATIVE TO start/ to build this string. This parameter
 *          is optional. */
char *dns_build_reply_string(char *data, int len, char *start, int *dist)
{
    char *ret_str = NULL;
    int str_size = 0;

    ret_str = malloc(128);
    memset(ret_str, 0x00, 128);
    
    dns_brs_helper(data, start, ret_str, &str_size, 0, 1);
    if (dist)
        *dist = str_size;
    
    return ret_str;    
}

/* Function that returns a string indicating what kind of resource
 * type matches the given code. DO NOT free the returned pointer,
 * it is a constant string... */
char *dns_type_to_str(uint16_t type)
{
    switch (type) {
      case DNS_QT_A:     return "A";
      case DNS_QT_NS:    return "NS";
      case DNS_QT_CNAME: return "CNAME";
      case DNS_QT_SOA:   return "SOA";
      case DNS_QT_PTR:   return "PTR";
      case DNS_QT_WKS:   return "WKS";
      case DNS_QT_AAAA:  return "AAAA";
      case DNS_QT_LOC:   return "LOC";
      case DNS_QT_HINFO: return "HINFO";
      case DNS_QT_MINFO: return "MINFO";
      case DNS_QT_MX:    return "MX";
      case DNS_QT_TXT:   return "TXT";
      case DNS_QT_AXFR:  return "AXFR";
      case DNS_QT_ANY:   return "ANY";
      case DNS_QT_RT:    return "RT";
      case DNS_QT_RP:    return "RP";
      case DNS_QT_AFSDB: return "AFSDB";
      case DNS_QT_A6:    return "A6";
      default:
        return "?";
    }
}

/* Function that returns a string indicating what class matches
 * a given class code. DO NOT free the returned pointer, it is a
 * constant string! */
char *dns_class_to_str(uint16_t class)
{
    switch (class) {
      case DNS_CLASS_INET:  return "INET";
      default:
        return "?";
    }
}

/* Function that returns a string describing one of the self-defined
 * error codes. If the error code is DNS_ERR_SEE_ERRNO, then strerror()
 * is called with the value of q->dns_errno. The returned string
 * MUST be free'd by the caller. */
char *dns_strerror(dnsq_t *q, int code)
{
    char *ret_str;
    
    if (code == DNS_ERR_SEE_ERRNO) {
        char *errno_str = strerror(q->dns_errno);
        ret_str = malloc(strlen(errno_str) + 10);
        snprintf(ret_str, strlen(errno_str) + 10, "(errno): %s",
                 errno_str);
        
        return ret_str;
    }
    
    switch (code) {
      case DNS_ERR_FORMAT_ERROR:
        return strdup("Badly formatted request string");
      case DNS_ERR_TOKEN_TOO_LONG:
        return strdup("A request token is too large (> 63)");
      case DNS_ERR_NULL_PARAM:
        return strdup("An illegal null value was given");
      case DNS_ERR_INVALID_QUERY:
        return strdup("Query structure is invalid");
      case DNS_ERR_PARAM_ERROR:
        return strdup("An illegal parameter value was given");
      case DNS_ERR_QUERY_TOO_LONG:
        return strdup("Query exceeds 512 bytes");
      case DNS_ERR_REPLY_TRUNCATED:
        return strdup("Server reply was truncated");
      case DNS_ERR_NO_SUCH_NAME:
        return strdup("No such name exists");
      case DNS_ERR_QUERY_REFUSED:
        return strdup("Server refused to handle this query");
      case DNS_ERR_BAD_FORMAT:
        return strdup("The sent query was badly formatted");
      case DNS_ERR_SERVER_FAILURE:
        return strdup("Server reports general failure");
      case DNS_ERR_NOT_IMPLEMENTED:
        return strdup("Server does not support requested query");
      case DNS_ERR_TIMED_OUT:
        return strdup("Timed out waiting for reply");
      default:
        return strdup("Unknown error code");
    };
}


dns.h
 /* DNS Query construction and reply parsing library. Note the following:
 *  - Currently only supports UDP
 *  - Zone transfers are currently unsupported (they require TCP) and
 *    attempting to add a zone transfer question will result in an
 *    DNS_INVALID_PARAM error
 * 
 * Written by Owen Klan  -  23rd August 2003 */

#ifndef __DNS_H_
#define __DNS_H_

/* Common include files for source files */
#include <sys/types.h>
#include <stdint.h>
#include <netinet/in.h>
#include <sys/socket.h>

/* Version symbols and macros to obtain them */
#define LIBDNS_VERSION   "0.1-alpha"
#define LIBDNS_MAJOR     0
#define LIBDNS_MINOR     1

#define dns_get_lib_version_str()     \
      (char *)LIBDNS_VERSION
#define dns_get_lib_version()         \
      (uint16_t)((LIBDNS_MAJOR << 8) | LIBDNS_MINOR)

/*
 * Declaration of LibDNS data structures
 */
/* Structure that forms the common header of the DNS query. */
struct dns_header {
    uint16_t dns_id;	       /* Identifier */
    uint16_t dns_flags;	       /* Flags word */
    uint16_t dns_num_quest;    /* # of questions */
    uint16_t dns_num_ans;      /* # of answer RRs */
    uint16_t dns_num_auth;     /* # of authority RRs */
    uint16_t dns_num_add;      /* # of additional RRs */
};

/*
 * The following structures are used to build lists of questions
 * and resource records.
 */
/* Question structure */
struct dns_question {
    char *       q_content;	       /* Question content */
    uint16_t     q_type;	       /* Question type */
    uint16_t     q_class;	       /* Question class */
    struct dns_question *next;	       /* Next question pointer */
};

/* Resource Record structure */
struct dns_rr {
    char *     rr_name;		       /* Domain name */
    uint16_t   rr_type; 	       /* RR type */
    uint16_t   rr_class;	       /* RR class */
    uint32_t   rr_ttl;  	       /* RR Time-to-live */
    uint16_t   rr_datalen;	       /* RR Data length */
    char *     rr_data;		       /* RR Data */
    struct dns_rr *next;	       /* Next RR in the list */
};

/* Query construction structure. */
struct dns_query {
    struct dns_header   header;
    struct dns_question *questions;    /* List of questions */
    struct dns_rr       *answers;      /* List of answers */
    struct dns_rr       *authority;    /* List of authority replies */
    struct dns_rr       *additional;   /* List of additional replies */
    int dns_errno;		       /* Local copy of last errno */
    char *dns_data;		       /* The reply data */
};
typedef struct dns_query dnsq_t;


/* 
 * Declaration of #define's for DNS fields.
 */
/* DNS Opcode values */
#define DNS_OP_QUERY          0x00     /* Query */
#define DNS_OP_INVQUERY       0x01     /* Inverse query */
#define DNS_OP_SERVSTAT       0x02     /* Server status */

/* DNS Query reply codes */
#define DNS_R_NO_ERROR        0	       /* No error */
#define DNS_R_FORMAT_ERROR    1	       /* Format error with query */
#define DNS_R_SERVER_FAILURE  2	       /* Server failure */
#define DNS_R_NAME_ERROR      3	       /* Name error */
#define DNS_R_NOT_IMPLEMENTED 4	       /* Query type not supported */
#define DNS_R_REFUSED         5	       /* Server refused to handle query */

/* Bits for 'recursive' option for dns_new_query() */
#define DNS_DO_RECURSE	      1        /* Enable recursion */
#define DNS_DONT_RECURSE      0	       /* Disable recursion */

/* DNS Query types */
#define DNS_QT_A              1	       /* IPv4 Address record */
#define DNS_QT_NS             2	       /* Name server */
#define DNS_QT_CNAME          5	       /* Canonical name (IP-hostname) */
#define DNS_QT_SOA            6        /* Start of zone-of-authority */
#define DNS_QT_WKS            11       /* Well Known Service description */
#define DNS_QT_PTR            12       /* Pointer record */
#define DNS_QT_HINFO          13       /* Host information */
#define DNS_QT_MINFO          14       /* Mail{box,list} information */
#define DNS_QT_MX             15       /* Mail exchange record */
#define DNS_QT_TXT            16       /* Text strings */
#define DNS_QT_RP             17       /* Responsible Person */
#define DNS_QT_AFSDB          18       /* AFS cell database server */
#define DNS_QT_RT             21       /* Route-through record */
#define DNS_QT_AAAA           28       /* IPv6 address record (RFC-1886) */
#define DNS_QT_LOC            29       /* Location information (RFC-1876) */
#define DNS_QT_SRV            33       /* Service information (RFC-2782) */
#define DNS_QT_A6             38       /* IPv6 A6 address record (RFC-2874) */
/* the following appear in query records only! */
#define DNS_QT_AXFR           252      /* Zone transfer */
#define DNS_QT_ANY            255      /* All records */

/* DNS Query classes */
#define DNS_CLASS_INET        1	       /* Inet address */
#define DNS_CLASS_CHAOS       3	       /* Chaos system */

/* Accessor routines declared as macros to get at elements of the dnsq_t
 * structure */
#define dns_get_data(q)               \
      (void *)((dnsq_t *)q)->dns_data
#define dns_get_question_list(q)      \
      (struct dns_rr *)((dnsq_t *)q)->questions
#define dns_get_answer_list(q)        \
      (struct dns_rr *)((dnsq_t *)q)->answers
#define dns_get_authority_list(q)     \
      (struct dns_rr *)((dnsq_t *)q)->authority
#define dns_get_additional_list(q)    \
      (struct dns_rr *)((dnsq_t *)q)->additional

/* Accessor routines declared as macros to get at elements of a resource
 * record structure */
#define dns_rr_get_data(r)        \
      (void *)((struct dns_rr *)r)->rr_data
#define dns_rr_get_data_len(r)    \
      (uint16_t)((struct dns_rr *)r)->rr_datalen
#define dns_rr_get_type(r)        \
      (uint16_t)((struct dns_rr *)r)->rr_type
#define dns_rr_get_class(r)       \
      (uint16_t)((struct dns_rr *)r)->rr_class
#define dns_rr_get_ttl(r)         \
      (uint32_t)((struct dns_rr *)r)->rr_ttl


/*
 * Macro that sets up the flags word as appropriate for a query.
 * o is the opcode field, r is the 'recursion desired' bit. The
 * Query-Response field will be set to 0 to indicate a query.
 * d is the dnsq_t structure to set the flags of.
 */
#define DNS_SET_FLAGS(d,o,r)                   \
    d->header.dns_flags = 0x0000;              \
    d->header.dns_flags |= ((o & 0x0F) << 11); \
    d->header.dns_flags |= (r << 8);

/* 
 * #define's for Libdns error codes
 */
#define DNS_ERR_FORMAT_ERROR            1  /* Format error of request string */
#define DNS_ERR_TOKEN_TOO_LONG          2  /* A token was longer than 63 */
#define DNS_ERR_SEE_ERRNO               3  /* Look at value of errno */
#define DNS_ERR_NULL_PARAM              4  /* A NULL parameter was given */
#define DNS_ERR_INVALID_QUERY           5  /* The given dnsq_t is invalid */
#define DNS_ERR_PARAM_ERROR		6  /* Function given invalid param */
#define DNS_ERR_QUERY_TOO_LONG          7  /* The whole query is too big
					    * (> 512 for UDP) */
#define DNS_ERR_REPLY_TRUNCATED         8  /* The response was truncated */
#define DNS_ERR_NO_SUCH_NAME            9  /* requested name doesn't exist */
#define DNS_ERR_QUERY_REFUSED          10  /* Server refused to handle query */
#define DNS_ERR_BAD_FORMAT             11  /* sent query was badly formatted */
#define DNS_ERR_SERVER_FAILURE         12  /* Server failure */
#define DNS_ERR_NOT_IMPLEMENTED        13  /* Query type not supported */
#define DNS_ERR_TIMED_OUT              14  /* Timed out waiting for reply */

/* Function prototypes */
/* From dns.c */
extern dnsq_t *dns_new_query(uint16_t id, int opcode, int recursive);
extern void dns_free_query(dnsq_t *q);
extern int dns_add_question(dnsq_t *q, char *host, uint16_t type,
			    uint16_t class);
extern int dns_send_query(dnsq_t *q, int s, struct sockaddr_in *addr,
			  int timeout);

/* From dns_str.c */
extern char *dns_convert_name(char *name, int *error);
extern char *dns_type_to_str(uint16_t type);
extern char *dns_class_to_str(uint16_t class);
extern char *dns_strerror(dnsq_t *q, int code);
extern char *dns_build_reply_string(char *data, int len,
				    char *start, int *dist);

/* Some helper macros, these are mostly for the library code, but could
 * be utilised by user apps. as well */
/* Macro that free's a pointer if it's non-NULL */
#define FREE_IF_VALID(p)     \
     { if (p)                \
	  free(p); }

/* Macro that is used by functions that return an error in a user
 * provided integer location. */
#define SET_ERROR(e)          \
    { if (error)              \
	*error = e; }

/* Macro to convert an unsigned short to network byte order, then
 * assign back into the variable */
#define FIX_FOR_NBO(s)    s = htons(s)

#endif				       /* __DNS_H_ */
